lib/Controller/TotpApiController.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ * @author Semih Serhat Karakaya <karakayasemi@itu.edu.tr>
+ *
+ * Two-factor TOTP
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\TwoFactor_Totp\Controller;
+
+use OC\OCS\Result;
+use OCA\TwoFactor_Totp\Exception\NoTotpSecretFoundException;
+use OCP\AppFramework\OCSController;
+use \OCP\IRequest;
+use \OCP\IUserManager;
+use OCA\TwoFactor_Totp\Service\ITotp;
+
+class TotpApiController extends OCSController {
+
+	/** @var ITotp */
+	private $totp;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	public function __construct($appName, IRequest $request, ITotp $totp, IUserManager $userManager) {
+		parent::__construct($appName, $request);
+		$this->totp = $totp;
+		$this->userManager = $userManager;
+	}
+
+	/**
+	 * @CORS
+	 * @NoCSRFRequired
+	 *
+	 * @param string $id
+	 * @param string $secret
+	 * @return Result
+	 */
+	public function validate($id, $secret) {
+		$user = $this->userManager->get($id);
+		if (!empty($user)) {
+			try {
+				return new Result(['result' => $this->totp->validateSecret($user, $secret)]);
+			} catch (NoTotpSecretFoundException $e) {
```

Comment:
I expect this to be a rare occurrence, so at least we should log the exception.

Reply:
Actually, not every user have a totp secret. We are not enforcing 2-Factor Totp usage. So, the frequency changes depending on the usage of the API. But, we can log the exception.